### PR TITLE
Various fixes

### PR DESCRIPTION
**Pull request by leemaguire**
* Fix iterator on `experimental::Results`
* Add `operator!=()` to collections.
* Add `set_schema_version(uint64_t)`
* Add `managed<std::vector<T*>>::push_back(const managed<T*>&)`
* Add `box<managed<V*>>::box& operator=(const managed<V*>& o)`
* Add `box<managed<V*>>::box& operator=(const managed<V>& o)`
* Add `managed<T*>::operator=(const managed<T>&)`
* Add `managed<T*>::operator=(const managed<T*>&)`
* Fix queries containing link object columns.
* Rename `managed<>::value()` to `managed<>::detach()`

### CODE REVIEW

File: src/cpprealm/experimental/link.hpp [hunk 1]
```diff
@@ -22,31 +22,57 @@ namespace realm {
...
+                    }
+                    return this->m_managed.m_obj.get_key() == rhs.m_obj->get_key();
+                }
                 bool operator ==(const managed<T>& rhs) const {
                     if (this->m_managed.m_realm != rhs.m_realm) {
                         return false;
                     }
-                    return this->m_managed.m_obj.get_table() == rhs.m_obj.get_table() &&
-                           this->m_managed.m_obj.get_key() == rhs.m_obj.get_key();
+                    return this->m_managed.m_obj.get_key() == rhs.m_obj.get_key();
```
Comment (jsflax) [inline src/cpprealm/experimental/link.hpp:35]: Why did this change? Comparing table was relevant code.

File: src/cpprealm/experimental/link.hpp [hunk 2]
```diff
@@ -22,31 +22,57 @@ namespace realm {
...
+
                 }
                 bool operator ==(const ref_type& rhs) const {
                     if (this->m_managed.m_realm != rhs.m_managed.m_realm) {
                         return false;
                     }
-                    return this->m_managed.m_obj.get_table() == rhs.m_managed.m_obj.get_table()
-                           && this->m_managed.m_obj.get_key() == rhs.m_managed.m_obj.get_key();
+                    return this->m_managed.m_obj.get_key() == rhs.m_managed.m_obj.get_key();
```
Comment (jsflax) [inline src/cpprealm/experimental/link.hpp:42]: Same as above.

File: src/cpprealm/experimental/macros.hpp [hunk 3]
```diff
@@ -307,41 +318,103 @@ rbool managed<std::optional<type>>::operator op(const std::optional<type>& rhs)
...
+                   std::apply([&](auto &&...ptr) { \
+                        std::apply([&](auto &&..._name) { \
+                        ((*this.*ptr).assign(&m_obj, &m_realm, m_obj.get_table().get_column_key(_name)), ...); \
+                    }, managed_pointers_names); \
+                }, managed_pointers());                                                                \
+            }  \
              return *this;\
-        } \
+        }                                                                                          \
+        static managed prepare_for_query(const internal::bridge::realm& r) {                       \
```
Comment (jsflax) [inline src/cpprealm/experimental/macros.hpp:404]: Can you give more detail for what these macro changes are?

File: src/cpprealm/experimental/managed_dictionary.hpp [hunk 4]
```diff
@@ -331,23 +340,35 @@ namespace realm::experimental {
             }
             return a.get_table() == b.get_table() && a.get_key() == b.get_key();
         }
+        bool operator!=(const V*& rhs) const {
+            return !this->operator==(rhs);
+        }
 
         bool operator==(const managed<V*> &rhs) const {
-            auto a = const_cast<box<V*> *>(this)->m_backing_map.get_object(this->m_key);
+           // const realm::experimental::box<realm::experimental::managed<realm::experimental::AllTypesObjectEmbedded *>> *' to 'box<realm::experimental::AllTypesObjectEmbedded *> *'
```
Comment (jsflax) [inline src/cpprealm/experimental/managed_dictionary.hpp:348]: Commented out code.

File: src/cpprealm/experimental/managed_dictionary.hpp [hunk 5]
```diff
@@ -429,8 +470,22 @@ namespace realm::experimental {
         using managed<std::map<std::string, T>>::managed_base::operator=;
 
         [[nodiscard]] std::map<std::string, T> value() const {
-            // unused
-            return std::map<std::string, T>();
+            if constexpr (std::is_pointer_v<T>) {
+                throw std::runtime_error("value() is not available on collections of managed objects. Access each object via subscript instead.");
```
Comment (jsflax) [inline src/cpprealm/experimental/managed_dictionary.hpp:474]: Why not?

File: src/cpprealm/experimental/managed_list.hpp [hunk 6]
```diff
@@ -117,8 +134,7 @@ namespace realm::experimental {
     template<typename T>
     struct managed<std::vector<T*>> : managed_base {
         [[nodiscard]] std::vector<T*> value() const {
-            // unused
-            return std::vector<T*>();
+            throw std::runtime_error("value() is not available on collections of managed objects. Access each object via subscript instead.");
```
Comment (jsflax) [inline src/cpprealm/experimental/managed_list.hpp:137]: Same question as above.

File: src/cpprealm/experimental/managed_numeric.hpp [hunk 7]
```diff
@@ -18,34 +18,96 @@ namespace realm::experimental {
...
+            m_obj->template set<int64_t>(this->m_key, old_val + 1);
         }
-        void operator-=(const int64_t& o) {
+        managed& operator-=(const int64_t& o) {
             auto old_val = m_obj->template get<int64_t>(m_key);
             m_obj->template set<int64_t>(this->m_key, old_val - o);
+            return *this;
         }
-        void operator--() {
+        void operator--(int) {
```
Comment (jsflax) [inline src/cpprealm/experimental/managed_numeric.hpp:103]: Why the change? This was for the iterator ifc.

File: src/cpprealm/experimental/macros.hpp (hunk 3 above)
Comment (leemaguire) [reply src/cpprealm/experimental/macros.hpp:404]: If you have a link property you need to propagate the `prepare_for_query` logic down the chain.

File: src/cpprealm/experimental/managed_dictionary.hpp (hunk 5 above)
Comment (leemaguire) [reply src/cpprealm/experimental/managed_dictionary.hpp:474]: `T` would be `MyObject*` where it needs to be `managed<MyObject*>`. Either we expose a different function for managed objects in collections or refactor value() to be able to return a different type.

File: src/cpprealm/experimental/managed_numeric.hpp (hunk 7 above)
Comment (leemaguire) [reply src/cpprealm/experimental/managed_numeric.hpp:103]: I've added back the prefix operators.
